[IRGen] Mix -profile-generate into module hash

beccadax opened the pull request:
When rebuilding a .o file, we hash the IR generated by IRGen and compare it to a hash embedded in the .o file. If they match, we simply skip asking LLVM to optimize and generate code. Certain flags that control our LLVM configuration are mixed into the hash, but the -profile-generate flag was not one of them. This usually wouldn’t matter because profiling would insert additional code into the IR, but in edge cases like empty files or files containing only protocol declarations, it could cause linker errors when profiling was turned off.

This change adds the `GenerateProfile` flag into the IR hash, ensuring that we always recompile, even if the IR was identical.

Fixes rdar://problem/54126622.

beccadax committed:
[IRGen] Mix -profile-generate into module hash

When rebuilding a .o file, we hash the IR generated by IRGen and compare it to a hash embedded in the .o file. If they match, we simply skip asking LLVM to optimize and generate code. Certain flags that control our LLVM configuration are mixed into the hash, but the -profile-generate flag was not one of them. This usually wouldn’t matter because profiling would insert additional code into the IR, but in edge cases like empty files or files containing only protocol declarations, it could cause linker errors when profiling was turned off.

This change adds the `GenerateProfile` flag into the IR hash, ensuring that we always recompile, even if the IR was identical.

Fixes rdar://problem/54126622.

brentdax commented:
@swift-ci please smoke test

gottesmm commented:
This is for @eeckstein

vedantk commented:
Thanks for working on this. To be on the safe side, should other types of instrumentation be included here as well (say, {a,t}san)?

eeckstein reviewed (approved):
lgtm!

gottesmm commented:
Question. Would it be possible to change this into some sort of exhaustive switch for Options? That way if someone adds a new option like this, they get a warning telling them that they may need to update this code.

gottesmm commented:
In other words, I am worried that the code as written will cause this same bug to occur again. If we had some sort of switch associated with it, at least we could get a compile time warning in such a case. This will cause the writer to at least think about... should this be used in the IR hash?

jrose-apple commented:
It's only required for options that don't affect the IR as generated, but arguably being forced to say that explicitly would be nice. You could do this with a .def file but that might feel heavyweight.

gottesmm commented:
@jrose-apple exactly. My fear is that someone will just add the option and not think about it. But if one at least has to explicitly do something, that can not happen.

brentdax commented:
@vedantk Good idea to check. `-sanitize=address` and `-sanitize=thread` emit different IR for an empty source file. `-sanitize=undefined` emits the same IR, but also the same assembly on x86_64 at least. `-sanitize=fuzzer` emits the same IR but different assembly, so that one should be mixed in too.

brentdax commented:
@gottesmm @jrose-apple I don't think the problem here is that someone forgot to update the function when they added an option; I think it's that someone didn't think about this edge case of an empty file, or didn't know about LLVM's behavior in this situation. `-profile-generate` doesn't usually need to be mixed into the flags because it will normally cause IR changes; only in files with no executable code does it behave incorrectly.

jrose-apple commented:
We could do that in +Asserts builds. We're only losing incrementality by doing so.

beccadax committed:
[NFC] Extract helper from performLLVM()

We do a particular dance to diagnose errors twice, and we’re about to do it three times.

beccadax committed:
Assert correctness of performLLVM() incremental builds

In assert builds, when performLLVM() would normally skip invoking LLVM because it believes it would generate the same code, it now generates code into a temporary file and compares it to the output. This should catch mistakes where Swift configures LLVM in ways which affect the output, but which the incremental logic doesn’t account for.

beccadax committed:
Mix fuzzer flag into IRGen module hash

brentdax commented:
This is probably the most complicated and expensive "assertion" I've ever written, but it's in and it works.

brentdax commented:
@swift-ci please test

jrose-apple commented on lib/IRGen/IRGen.cpp:
```diff
@@ -561,6 +591,45 @@ bool swift::performLLVM(IRGenOptions &Opts, DiagnosticEngine *Diags,
+                   << "\n  New file: " << OutputFilename << "\n";
+      if (DiagMutex)
+        DiagMutex->unlock();
+      llvm_unreachable(
```
Nitpick: This should use `llvm::report_fatal_error`.

jrose-apple commented on lib/IRGen/IRGen.cpp:
```diff
@@ -561,6 +591,45 @@ bool swift::performLLVM(IRGenOptions &Opts, DiagnosticEngine *Diags,
+                                          OriginalOutputFilename,
+                                          /*tolerances=*/0, 0, &message)) {
+      if (DiagMutex)
+        DiagMutex->lock();
```
I don't think you need to take the mutex here (or below); you're just going to crash anyway, and the temporary file doesn't affect other threads.

brentdax replied:
The mutex is to keep several threads from writing messages at the same time (or at least that's my reading of why it's done elsewhere in the file), but I think I'll simplify the output a bit anyway.

jrose-apple commented on lib/IRGen/IRGen.cpp:
```diff
@@ -561,6 +591,45 @@ bool swift::performLLVM(IRGenOptions &Opts, DiagnosticEngine *Diags,
+    std::string message;
+    if (0 != llvm::DiffFilesWithTolerance(OutputFilename,
+                                          OriginalOutputFilename,
+                                          /*tolerances=*/0, 0, &message)) {
```
This seems like overkill. Check out `swift::moveFileIfDifferent`.

jrose-apple commented on lib/IRGen/IRGen.cpp:
```diff
@@ -561,6 +591,45 @@ bool swift::performLLVM(IRGenOptions &Opts, DiagnosticEngine *Diags,
+                     << EC.message() << "\n";
+        if (DiagMutex)
+          DiagMutex->unlock();
+        llvm_unreachable("Unexpected error deleting temp file");
```
Here too…or we could just ignore it.

jrose-apple reviewed (commented):
Nice work. cc @eeckstein

beccadax committed:
[NFC] Extract helper from moveFileIfDifferent()

This separates the “do these two files have the same contents?” logic from the “move or delete” logic in `moveFileIfDifferent()`, creating a useful helper function. It also ties the special-case behavior for the `destination` parameter to a flag, since we have a use where we won’t want that.

jrose-apple commented on include/swift/AST/IRGenOptions.h:
```diff
@@ -259,6 +259,8 @@ class IRGenOptions {
     Hash = (Hash << 1) | DisableLLVMOptzns;
     Hash = (Hash << 1) | DisableSwiftSpecificLLVMOptzns;
+    Hash = (Hash << 1) | GenerateProfile;
+    Hash = (Hash << 1) | Sanitizers.contains(SanitizerKind::Fuzzer);
```
Honestly we might as well mix in the whole thing. It's not just about empty files; it's also about *non-*empty files that only change due to LLVM passes, and I'm not sure I trust that all the other sanitizers don't need this.

jrose-apple commented on test/IRGen/Inputs/empty.swift:
```diff
@@ -0,0 +1 @@
+// This space intentionally left blank.
```
Nitpick: There's already a test/Inputs/empty.swift you can use.

jrose-apple commented:
Thanks for the refactor!

brentdax commented:
@swift-ci please smoke test

beccadax committed:
[NFC] Hash LLVM options more robustly

Replaces getLLVMCodeGenOptionsHash(), which combined a bunch of individual bits into a string, with writeLLVMCodeGenOptionsTo(), which writes each one separately into a raw_ostream.

This is intended to be make the hashing more future-proof. The current design needs to know exactly how many bits each of the values needs; if any of the values grew and you forgot to update this function, its bits would interfere with those of an earlier value in the hash. This new design is expected to be slightly slower, but more robust to future change in the compiler.

beccadax committed:
[probably NFC] Mix all sanitizers into module hash

Mixes the state of all sanitizers, not just the fuzzer, into the module hash used to decide whether to skip LLVM codegen. I don’t actually know of a case where one of the other sanitizers will generate identical IR and different machine code, but being defensive costs us very little.

beccadax committed:
[NFC] Remove redundant empty test case file

brentdax commented:
@jrose-apple I mixed in all of the sanitizers, and also redesigned the hash slightly to avoid any issues when we add more sanitizers/optimization modes/values to these fields in general.

brentdax commented:
@swift-ci please smoke test
